core/features/feature_engineering.py

```diff
@@ -0,0 +1,400 @@
+import numpy as np
+import polars as pl
+
+from core.utils.general_utils import create_logger
+
+logger = create_logger(__name__)
+
+
+def interpolate_population_density(
+    df: pl.DataFrame, pixel_resolution: str, year_intervals: list[tuple[int, int]]
+) -> pl.DataFrame:
+    """
+    Interpolates population density values in between available years, as well
+    as going back to the first year in the PREDICTS database. The growth rate
+    for each interval is based on the start and end year of that intervale.
+    The calculation assumes an exponential growth curve.
+
+    Args:
+        df: Dataframe containing one site id column ('SSBS') and a number of
+            year columns. These should match with the year intervals.
+        pixel_resolution: The resolutions of this population data.
+        year_interval: The year intervals that population data needs to be
+            interpolated between. The first year in PREDICTS is 1984 and the
+            last is 2018. Population data is available 2000, 2005, 2010, 2015
+            and 2020.
+
+    Returns:
+        df: Original df with additional columns for all interpolated years.
+    """
+    logger.info(f"Population density interpolation for resolution {pixel_resolution}.")
+
+    # Helper function to calculate growth rate between two years
+    def _calculate_growth_rate(
+        df: pl.DataFrame, start_year: int, end_year: int
+    ) -> pl.Series:
+        rates = np.log(
+            df.get_column(str(end_year)) / df.get_column(str(start_year))
+        ) / (end_year - start_year)
+
+        # If there are NaN or inf values, fill with zeros
+        logger.warning(
+            f"NaN or inf values in growth rate for {start_year}-{end_year}. \
+                Setting growth rates to 0."
+        )
+        if rates.is_nan().any() or rates.is_infinite().any():
+            rates = (
+                pl.when(rates.is_nan() | rates.is_infinite()).then(0).otherwise(rates)
+            )
+
+        return rates
+
+    # Extrapolate back to 1984 using the growth rate from 2000 to 2005
+    r_backwards = _calculate_growth_rate(df, year_intervals[1][0], year_intervals[1][1])
+    df = df.with_columns(
+        (
+            df[str(year_intervals[1][0])]
+            * np.exp(r_backwards * (year_intervals[0][0] - year_intervals[0][1]))
+        ).alias("1984")
+    )
+
+    # Loop through each interval to calculate growth rates and interpolate
+    for start_year, end_year in year_intervals:
+        r = _calculate_growth_rate(df, start_year, end_year)
+        for year in range(start_year, end_year + 1):
+            if year not in df.columns:
+                df = df.with_columns(
+                    (df[str(start_year)] * np.exp(r * (year - start_year))).alias(
+                        str(year)
+                    )
+                )
```

Comment:
I wonder if `df = df.with_columns` is redundant? It looks like `df.with_columns` does an in place operation, so you might not need to keep writing over `df`?

Reply:
https://docs.pola.rs/py-polars/html/reference/dataframe/api/polars.DataFrame.with_columns.html

Reply:
I think you are right. I believe that's an old `pandas` habit to avoid the "setting with copy" warning. I don't think it's needed in `polars` since copying and memory works differently. I'm guilty of this in many place - will add to the backlog.